[libc++] Rejects-valid on unexpected(1) == unexpected(1L)

Minimal reproducer:
```
#include <expected>
static_assert(std::unexpected(1) == std::unexpected(1L));
```

Output:
```
__expected/unexpected.h:111:31: error: '__unex_' is a private member of 'std::unexpected<long>'
  111 |     return __x.__unex_ == __y.__unex_;
      |                               ^
```

Godbolt: https://godbolt.org/z/a8rPnKz1Y

The fix is to use `__y.error()` instead of `__y.__unex_`.

This was originally discovered in https://github.com/Quuxplusone/llvm-project/issues/40.

--- a/libcxx/include/__expected/unexpected.h
+++ b/libcxx/include/__expected/unexpected.h
@@ -108,7 +108,7 @@ public:
 
   template <class _Err2>
   _LIBCPP_HIDE_FROM_ABI friend constexpr bool operator==(const unexpected& __x, const unexpected<_Err2>& __y) {
-    return __x.__unex_ == __y.__unex_;
+    return __x.__unex_ == __y.error();
   }
 
 private:

--- a/libcxx/test/std/utilities/expected/expected.unexpected/equality.pass.cpp
+++ b/libcxx/test/std/utilities/expected/expected.unexpected/equality.pass.cpp
@@ -20,18 +20,29 @@
 #include <expected>
 #include <utility>
 
-struct Error{
+struct Error {
   int i;
   friend constexpr bool operator==(const Error&, const Error&) = default;
+  friend constexpr bool operator==(const Error& lhs, int rhs) noexcept { return lhs.i == rhs; }
 };
 
 constexpr bool test() {
   std::unexpected<Error> unex1(Error{2});
   std::unexpected<Error> unex2(Error{3});
   std::unexpected<Error> unex3(Error{2});
+
   assert(unex1 == unex3);
   assert(unex1 != unex2);
   assert(unex2 != unex3);
+
+  std::unexpected<int> unex_i1(1);
+  std::unexpected<int> unex_i2(2);
+
+  assert(unex1 != unex_i1);
+  assert(unex_i1 != unex1);
+  assert(unex1 == unex_i2);
+  assert(unex_i2 == unex1);
+
   return true;
 }
 

